### ISSUE TITLE
conch tests make heavy use of deprecated assert* free functions

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2112]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2112|
|Type|defect|
|Created|2006-09-20 21:55:14Z|
|Branch|https://github.com/twisted/twisted/tree/free-asserts-2112-4|

<details><summary>Searchable metadata</summary>

```
trac-id__2112 2112
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_free_asserts_2112_4 branches/free-asserts-2112-4
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__soc2007 soc2007
time__1158789314000000 1158789314000000
changetime__1300330131000000 1300330131000000
version__None None
owner__ 
cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_2"></a>|@dreid removed owner|
|-|-|

branched to [branches/free-assert-2112](branches/free-assert-2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

While this is certainly an improvement (deprecation warnings are no more), I'm not sure it's a good idea.  Going from the version of the tests in this branch to _actually_ correctly written tests may be harder than getting to that destination from the version of these tests in trunk.

What should really be going on here is that the various classes involved should be mocked so that the test method itself can inspect them and make assertions about their state.  Having a reference to the test case in all of the places currently making assertions doesn't make any progress towards this goal.

Leaving this up for review.  I think it's lateral movement, but someone else may see it some other way.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_4"></a>|@dreid commented|
|-|-|

It's a valid point that it may indeed be more difficult to go to properly factored tests since there are no more deprecation warnings to motivate people.  However I don't think I'll be able to put that amount of work into it anytime soon. Plus i don't want to get suckered into conch maintainership or anything like that.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

I'm with exarkun on this.  I also don't want to be suckered into maintaining conch, but I might be able to fix the tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I guess someone else is going to take another whack at this in a less lazy way.  De-review-ifying.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to z3p|
|-|-|

Assigning to our perpetually prodigal developer, now that he has a trac account again...

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|z3p commented|
|-|-|

Tagging soc2007

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [30942]) Branching to 'free-asserts-2112-4'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [30943]) Remove the assert from ConchTestRealm by just making the requestAvatar method fail if the username is wrong.

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [30944]) Check the key and fingerprint in verifyHostKey and return a Failure if they do not match, rather than making an assertion and raising an AssertionError

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [30945]) If the channel open fails, log the reason it failed instead of using unittest.fail to raise an AssertionError; either way the test will fail (and neither is as good as failing a waiting Deferred that has been handed to trial).

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [30946]) Invert the assertion responsibility in test_ourServerOurClient for the scope of SSHTestSubsystemChannel, removing the use of deprecated free assertion functions

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [30947]) Parameterize the channel that is used in the one loopback test method, getting ready to add a bunch more test methods each which exercise one of these operations

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [30948]) This is not necessary for the test to operate

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [30949]) Re-introduce the pty-req/shell tests in the new-style

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [30950]) port failed exec and false channel test code

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [30952]) Port the max packet test code

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [30953]) Port the rest of the channel operation test code

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [30954]) Okay, just kidding, one more

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [30955]) Delete some more dead code, keep track of the avatar to make assertions about server-side state in unit tests

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [30956]) Remove some asserts from the server-side test spies

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [30957]) Replace free assertions with simple checks and appropriate error returns in various cred objects

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

(In [30962]) Log an error if the disconnect is for an unexpected reason, instead of making an assertion

refs [#2112](https://github.com/twisted/twisted/issues/2112)

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_28"></a>|@lvh commented|
|-|-|

Looks good to me and the buildbots don't fall apart (except unrelated intermittent failure on win7).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set status to `closed`|
|-|-|

(In [31213]) Merge free-asserts-2112-4 

Author: exarkun
Reviewer: lvh
Fixes: [#2112](https://github.com/twisted/twisted/issues/2112)

Refactor some Conch test code to split up big tests into many smaller
tests and to stop using deprecate free assertion functions from the
unittest module.
